File: C_version/src/i2c.c
#include "i2c.h"
#include "usart.h"

void i2c_master_setup(void) {
    // SCL_FREQ F_CPU/(16 + 2*(TWBR_VALUE)*PRESCALER)
    // Set I2C frequency to 400kHz
    TWSR = 0x00; // prescaler = 1
    TWBR = 0x0C; // 0x0C = 12 
    TWCR |= (1 << TWEN); // Turn on TWI/I2C module  
}

void i2c_master_start(void) {
    // Send start bit, enable interrupt, enable TWI  
    TWCR = (1<<TWINT)|(1<<TWSTA)|(1<<TWEN); 
    while ((TWCR & (1 << TWSTA)) == 0); // Wait until start bit sent
    if (i2c_master_get_status() != START_SUCCESS) { panic_msg("Paniced at start!"); }
}

void i2c_master_stop(void) {
    // Send stop bit, enable interrupts, enable TWI
    TWCR = (1<<TWINT)|(1<<TWSTO)|(1<<TWEN); 
}

void i2c_master_send(uint8_t data_byte) {
    TWDR = data_byte; // Store data in data register
    TWCR = (1 << TWINT) | (1 << TWEN);
    while ((TWCR & (1<<TWINT)) == 0); // Wait for transmission
    if (i2c_master_get_status() != SLA_ACK_SUCCESS) { panic_msg("Paniced at send!"); }
}

uint8_t i2c_master_read_ack(void) {
    TWCR = (1<<TWINT)|(1<<TWEN)|(1<<TWEA); 
    while ((TWCR & (1<<TWINT)) == 0); // Wait for transmission
    return TWDR; 
}

uint8_t i2c_master_read_nack(void) {
    TWCR = (1<<TWINT)|(1<<TWEN); 
    while ((TWCR & (1<<TWINT)) == 0); // Wait for transmission
    return TWDR; 
}

uint8_t i2c_master_get_status(void) {
    uint8_t status;
    status = TWSR & 0b11111000; // Mask status
    return status;
}

void i2c_write_byte(uint8_t Wadd, uint8_t reg, uint8_t value){
    i2c_master_start();       // Start bit
    // if (i2c_master_get_status() != START_SUCCESS) { panic(); }
    
    i2c_master_send(Wadd);    // Send address for write
    // if (i2c_master_get_status() != SLA_ACK_SUCCESS) { panic(); }
    
    i2c_master_send(reg);     // Send data - which register?
    // if (i2c_master_get_status() != DATA_ACK_SUCCESS) { panic(); }
   
    i2c_master_send(value);   // Send data - what value?
    // if (i2c_master_get_status() != DATA_ACK_SUCCESS) { panic(); }
  
    i2c_master_stop();        // Stop bit   

}

// maybe should be i16?
// should it be nack?
uint8_t i2c_read_byte(uint8_t Wadd, uint8_t Radd, uint8_t reg){
    // uint8_t Radd = Wadd | 0b00000001;
    uint8_t recv;
   
    i2c_master_start();             // Send start bit
    // if (i2c_master_get_status() != START_SUCCESS) { panic(); }
    
    i2c_master_send(Wadd);          // Send write address
    // if (i2c_master_get_status() != SLA_ACK_SUCCESS) { panic(); }
   
    i2c_master_send(reg);           // Send the register we want to read from 
    // if (i2c_master_get_status() != DATA_ACK_SUCCESS) { panic(); }
  
    i2c_master_start();             // Restart
    // if ((i2c_master_get_status()) != RESTART_SUCCESS) { panic(); }
 
    i2c_master_send(Radd);          // Send read address

    recv = i2c_master_read_ack();   // Get received value, send ack

    i2c_master_stop();              // Stop bit
    
    return recv;
}

void i2c_read_multiple(uint8_t Wadd, uint8_t Radd, uint8_t reg, uint8_t *raw, int len){
    // uint8_t Radd = Wadd | 0b00000001; // Change the last bit to 1 for read address 
    int i;
    i2c_master_start();       // Send start bit
    // if (i2c_master_get_status() != START_SUCCESS) { panic(); }
    
    i2c_master_send(Wadd);    // Send write address
    // if (i2c_master_get_status() != SLA_ACK_SUCCESS) { panic(); }
    
    i2c_master_send(reg);     // Send the register we want to read from 
    // if (i2c_master_get_status() != DATA_ACK_SUCCESS) { panic(); }
    
    i2c_master_start();     // Restart
    // if ((i2c_master_get_status()) != RESTART_SUCCESS) { panic(); }
    i2c_master_send(Radd);    // Send read address
    
    for (i = 0; i < len; i++){
        if (i == (len-1)){  // break out at the last one 
            raw[i] = i2c_master_read_ack(); 
            break;
        }
        else{
            raw[i] = i2c_master_read_nack();
        }
         
    }
    i2c_master_stop();        // Stop bit

}

File: examples/unsafe_usart/nano_hal/src/i2c.rs
// https://embedds.com/programming-avr-i2c-interface/

use crate::meta::*;
use crate::atmega328p::*;
use core::ptr::read_volatile;
use core::ptr::write_volatile;

// pub const SCL_PRE: u8 = 1;
// pub const SCL_FREQ: u32 = F_CPU / 16 + 2*(TWBR) * (SCL_PRE)

pub unsafe fn init() {
    // Set SCL to 400kHz
    write_volatile(TWSR, 0x00);
    write_volatile(TWBR, 0x0C); // 0x0C = 12, 16000000/(16+(2*12)) = 400 kHz
    
    // Enable TWI
    set_bit(TWCR, *TWEN, true)
}

pub unsafe fn start() {
    set_bit(TWCR, *TWINT, true);
    set_bit(TWCR, *TWSTA, true);
    set_bit(TWCR, *TWEN, true);

    while ( read_volatile(TWCR) & _BV(*TWINT) ) == 0 {
        // wait until TWINT resets to zero
    }
}

pub unsafe fn stop() {
    set_bit(TWCR, *TWINT, true);
    set_bit(TWCR, *TWSTO, true);
    set_bit(TWCR, *TWEN, true);
}

pub unsafe fn write(data: u8) {
    write_volatile(TWDR, data);
    set_bit(TWCR, *TWINT, true);
    set_bit(TWCR, *TWEN, true);

    while ( read_volatile(TWCR) & _BV(*TWINT) ) == 0 {
        // wait until TWINT resets to zero
    }
}

pub unsafe fn read_ACK() -> u8 {
    set_bit(TWCR, *TWINT, true);
    set_bit(TWCR, *TWEN, true);
    set_bit(TWCR, *TWEA, true);
    while ( read_volatile(TWCR) & _BV(*TWINT) ) == 0 {
        // wait until TWINT resets to zero
    }
    read_volatile(TWDR)
}

pub unsafe fn read_NACK() -> u8 {
    set_bit(TWCR, *TWINT, true);
    set_bit(TWCR, *TWEN, true);
    while ( read_volatile(TWCR) & _BV(*TWINT) ) == 0 {
        // wait until TWINT resets to zero
    }
    read_volatile(TWDR)
}

pub unsafe fn get_status() -> u8 {
    let status: u8;
    
    // Mask status
    status = read_volatile(TWSR) & 0xF8;
    status
}

pub unsafe fn set_pin(Wadd: u8, reg: u8, value: u8) {
    start();
    write(Wadd);
    write(reg);
    write(value);
    stop();
}

// pub unsafe fn read_pin(Wadd: u8, reg: u8) -> u8 {
    // let Radd: u8 = Wadd | 0b01000001;
    // let mut recv: u8;
//
    // start();
    // send(Wadd);
    // send(reg);
//
//
//
// }
//


// void setpin(unsigned char wadd, unsigned char reg, unsigned char value){
    // i2c_master_start();       // Start bit
    // i2c_master_send(Wadd); // Send address for write
    // i2c_master_send(reg);     // Send data - which register?
    // i2c_master_send(value);     // Send data - what value?
    // i2c_master_stop();        // Stop bit
// }









